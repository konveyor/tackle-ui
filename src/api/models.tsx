export interface PageQuery {
  page: number;
  perPage: number;
}

export interface SortByQuery {
  index: number;
  direction: "asc" | "desc";
}

export interface PageRepresentation<T> {
  meta: Meta;
  data: T[];
}

export interface Meta {
  count: number;
}

// Controls

export interface BusinessService {
  id?: number;
  name: string;
  description?: string;
  owner?: Stakeholder;
}

export interface Stakeholder {
  id?: number;
  displayName: string;
  email: string;
  jobFunction?: JobFunction;
  stakeholderGroups?: StakeholderGroup[];
}

export interface StakeholderGroup {
  id?: number;
  name: string;
  description: string;
  stakeholders?: Stakeholder[];
}

export interface JobFunction {
  id?: number;
  role: string;
}

export interface TagType {
  id?: number;
  name: string;
  rank?: number;
  colour?: string;
  tags?: Tag[];
}

export interface Tag {
  id?: number;
  name: string;
  tagType?: TagType;
}

// Application inventory

export interface Application {
  id?: number;
  name: string;
  description?: string;
  comments?: string;
  businessService?: string;
  tags?: string[];
  review?: Review;
}

export interface Review {
  id?: number;
  proposedAction: string;
  effortEstimate: string;
  businessCriticality: number;
  workPriority: number;
  comments?: string;
  application?: Application;
}

export interface ApplicationDependency {
  id?: number;
  from: Application;
  to: Application;
}

// Pathfinder

export type AssessmentStatus = "EMPTY" | "STARTED" | "COMPLETE";

export interface Assessment {
  id?: number;
  applicationId: number;
  status: "EMPTY" | "STARTED" | "COMPLETE";
  stakeholders?: number[];
  stakeholderGroups?: number[];
}

export interface Assessment {
  id?: number;
  applicationId: number;
  status: AssessmentStatus;
  stakeholders?: number[];
  stakeholderGroups?: number[];
  questionnaire: Questionnaire;
}

export interface Questionnaire {
  categories: QuestionnaireCategory[];
}

export interface QuestionnaireCategory {
  id: number;
  order: number;
  title?: string;
  comment?: string;
  questions: Question[];
}

export interface Question {
  id: number;
  order: number;
  question: string;
  description: string;
  options: QuestionOption[];
}

export interface QuestionOption {
  id: number;
  order: number;
  option: string;
  checked: boolean;
}

// Pagination

export interface BusinessServicePage {
  _embedded: {
    "business-service": BusinessService[];
  };
  total_count: number;
}

export interface StakeholderPage {
  _embedded: {
    stakeholder: Stakeholder[];
  };
  total_count: number;
}

export interface StakeholderGroupPage {
  _embedded: {
    "stakeholder-group": StakeholderGroup[];
  };
  total_count: number;
}

export interface JobFunctionPage {
  _embedded: {
    "job-function": JobFunction[];
  };
  total_count: number;
}

export interface TagTypePage {
  _embedded: {
    "tag-type": TagType[];
  };
  total_count: number;
}

export interface TagTypePage {
  _embedded: {
    "tag-type": TagType[];
  };
  total_count: number;
}

export interface ApplicationPage {
  _embedded: {
    application: Application[];
  };
  total_count: number;
}

export interface ApplicationDependencyPage {
  _embedded: {
    "applications-dependency": ApplicationDependency[];
  };
  total_count: number;
}
