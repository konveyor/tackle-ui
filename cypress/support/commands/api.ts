/// <reference types="cypress" />

import { buildHeadersWithHAL_JSON } from "./utils";

const sizeQueryParam = "size=1000";
const controlsBaseUrl = "/api/controls";
const applicationInventoryBaseUrl = "/api/application-inventory";

export type ResourceType =
  | "JobFunction"
  | "Stakeholder"
  | "StakeholderGroup"
  | "BusinessService"
  | "TagType"
  | "Tag"
  | "Application"
  | "ApplicationsDependency";

const allResourceTypes: ResourceType[] = [
  "JobFunction",
  "Stakeholder",
  "StakeholderGroup",
  "BusinessService",
  "TagType",
  "Tag",
  "Application",
  "ApplicationsDependency",
];

type ResourceTypeList = {
  [key in ResourceType]: {
    baseUrl: string;
    _embeddedField: string;
  };
};

const resourceTypeList: ResourceTypeList = {
  JobFunction: {
    baseUrl: `${controlsBaseUrl}/job-function`,
    _embeddedField: "job-function",
  },
  Stakeholder: {
    baseUrl: `${controlsBaseUrl}/stakeholder`,
    _embeddedField: "stakeholder",
  },
  StakeholderGroup: {
    baseUrl: `${controlsBaseUrl}/stakeholder-group`,
    _embeddedField: "stakeholder-group",
  },
  BusinessService: {
    baseUrl: `${controlsBaseUrl}/business-service`,
    _embeddedField: "business-service",
  },
  TagType: {
    baseUrl: `${controlsBaseUrl}/tag-type`,
    _embeddedField: "tag-type",
  },
  Tag: {
    baseUrl: `${controlsBaseUrl}/tag`,
    _embeddedField: "tag",
  },
  Application: {
    baseUrl: `${applicationInventoryBaseUrl}/application`,
    _embeddedField: "application",
  },
  ApplicationsDependency: {
    baseUrl: `${applicationInventoryBaseUrl}/applications-dependency`,
    _embeddedField: "applications-dependency",
  },
};

const create = (headers: any, url: string, payload: any) => {
  cy.request({
    method: "POST",
    headers: headers,
    body: payload,
    url: url,
  }).its("body");
};

const deleteAll = (headers: any, url: string, embeddedField: string) => {
  cy.request({
    method: "GET",
    headers: headers,
    url: `${url}?${sizeQueryParam}`,
  })
    .then((response) => response.body._embedded[embeddedField])
    .each((item: any) =>
      cy.request({
        method: "DELETE",
        headers: headers,
        url: `${url}/${item.id}`,
      })
    );
};

// Commands

declare global {
  namespace Cypress {
    interface Chainable<Subject> {
      api_create(
        tokens: KcTokens,
        resource: ResourceType,
        payload: any
      ): Chainable<any>;
      api_clean(tokens: KcTokens, resource?: ResourceType): void;
    }
  }
}

Cypress.Commands.add(
  "api_create",
  (tokens: KcTokens, resource: ResourceType, payload: any) => {
    const headers = buildHeadersWithHAL_JSON(tokens);
    const type = resourceTypeList[resource];

    create(headers, type.baseUrl, payload);
  }
);

Cypress.Commands.add(
  "api_clean",
  (tokens: KcTokens, resource?: ResourceType) => {
    const headers = buildHeadersWithHAL_JSON(tokens);

    if (resource) {
      const type = resourceTypeList[resource];
      deleteAll(headers, type.baseUrl, type._embeddedField);
    } else {
      allResourceTypes.forEach((e) => {
        const type = resourceTypeList[e];
        deleteAll(headers, type.baseUrl, type._embeddedField);
      });
    }
  }
);

export {};
