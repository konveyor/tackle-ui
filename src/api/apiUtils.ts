import {
  Application,
  ApplicationDependency,
  ApplicationDependencyPage,
  ApplicationPage,
  BusinessService,
  BusinessServicePage,
  JobFunction,
  JobFunctionPage,
  PageRepresentation,
  Stakeholder,
  StakeholderGroup,
  StakeholderGroupPage,
  StakeholderPage,
  TagType,
  TagTypePage,
} from "./models";
import {
  ApplicationSortBy,
  BusinessServiceSortBy,
  getApplicationDependencies,
  getApplications,
  getBusinessServices,
  getJobFunctions,
  getStakeholderGroups,
  getStakeholders,
  getTagTypes,
  JobFunctionSortBy,
  StakeholderGroupSortBy,
  StakeholderSortBy,
  TagTypeSortBy,
  TagTypeSortByQuery,
} from "./rest";

export const getAllBusinessServices = () => {
  return getBusinessServices(
    {},
    { page: 1, perPage: 1000 },
    { field: BusinessServiceSortBy.NAME }
  );
};

export const getAllStakeholders = () => {
  return getStakeholders(
    {},
    { page: 1, perPage: 1000 },
    { field: StakeholderSortBy.DISPLAY_NAME }
  );
};

export const getAllStakeholderGroups = () => {
  return getStakeholderGroups(
    {},
    { page: 1, perPage: 1000 },
    { field: StakeholderGroupSortBy.NAME }
  );
};

export const getAllJobFunctions = () => {
  return getJobFunctions(
    {},
    { page: 1, perPage: 1000 },
    { field: JobFunctionSortBy.ROLE }
  );
};

export const getAllTagTypes = (
  sortBy: TagTypeSortByQuery = { field: TagTypeSortBy.NAME }
) => {
  return getTagTypes({}, { page: 1, perPage: 1000 }, sortBy);
};

export const getAllApplications = () => {
  return getApplications(
    {},
    { page: 1, perPage: 1000 },
    { field: ApplicationSortBy.NAME }
  );
};

export const getAllApplicationDependencies = (filters: {
  from?: string[];
  to?: string[];
}) => {
  return getApplicationDependencies(filters, { page: 1, perPage: 1000 });
};

//

export const stakeholderPageMapper = (
  page: StakeholderPage
): PageRepresentation<Stakeholder> => ({
  meta: { count: page.total_count },
  data: page._embedded.stakeholder,
});

export const stakeholderGroupPageMapper = (
  page: StakeholderGroupPage
): PageRepresentation<StakeholderGroup> => ({
  meta: { count: page.total_count },
  data: page._embedded["stakeholder-group"],
});

export const bussinessServicePageMapper = (
  page: BusinessServicePage
): PageRepresentation<BusinessService> => ({
  meta: { count: page.total_count },
  data: page._embedded["business-service"],
});

export const jobFunctionPageMapper = (
  page: JobFunctionPage
): PageRepresentation<JobFunction> => ({
  meta: { count: page.total_count },
  data: page._embedded["job-function"],
});

export const tagTypePageMapper = (
  page: TagTypePage
): PageRepresentation<TagType> => ({
  meta: { count: page.total_count },
  data: page._embedded["tag-type"],
});

//

export const applicationPageMapper = (
  page: ApplicationPage
): PageRepresentation<Application> => ({
  meta: { count: page.total_count },
  data: page._embedded.application,
});

export const applicationDependencyPageMapper = (
  page: ApplicationDependencyPage
): PageRepresentation<ApplicationDependency> => ({
  meta: { count: page.total_count },
  data: page._embedded["applications-dependency"],
});
