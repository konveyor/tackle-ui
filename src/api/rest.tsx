import { AxiosPromise } from "axios";
import { APIClient } from "axios-config";

import {
  PageQuery,
  BusinessService,
  BusinessServicePage,
  Stakeholder,
  StakeholderPage,
  StakeholderGroup,
  StakeholderGroupPage,
  JobFunctionPage,
  ApplicationPage,
  Application,
  Assessment,
  JobFunction,
  ApplicationDependencyPage,
  ApplicationDependency,
} from "./models";

export const CONTROLS_BASE_URL = "controls";
export const APP_INVENTORY_BASE_URL = "application-inventory";
export const PATHFINDER_BASE_URL = "pathfinder";

export const BUSINESS_SERVICES = CONTROLS_BASE_URL + "/business-service";
export const STAKEHOLDERS = CONTROLS_BASE_URL + "/stakeholder";
export const STAKEHOLDER_GROUPS = CONTROLS_BASE_URL + "/stakeholder-group";
export const JOB_FUNCTIONS = CONTROLS_BASE_URL + "/job-function";

export const APPLICATIONS = APP_INVENTORY_BASE_URL + "/application";
export const APPLICATION_DEPENDENCY =
  APP_INVENTORY_BASE_URL + "/applications-dependency";

export const ASSESSMENTS = PATHFINDER_BASE_URL + "/assessments";

const headers = { Accept: "application/hal+json" };

type Direction = "asc" | "desc";

const buildQuery = (params: any) => {
  const query: string[] = [];

  Object.keys(params).forEach((key) => {
    const value = (params as any)[key];

    if (value !== undefined && value !== null) {
      let queryParamValues: string[] = [];
      if (Array.isArray(value)) {
        queryParamValues = value;
      } else {
        queryParamValues = [value];
      }
      queryParamValues.forEach((v) => query.push(`${key}=${v}`));
    }
  });

  return query;
};

// Business services

export enum BusinessServiceSortBy {
  NAME,
  OWNER,
}
export interface BusinessServiceSortByQuery {
  field: BusinessServiceSortBy;
  direction?: Direction;
}

export const getBusinessServices = (
  filters: {
    name?: string[];
    description?: string[];
    owner?: string[];
  },
  pagination: PageQuery,
  sortBy?: BusinessServiceSortByQuery
): AxiosPromise<BusinessServicePage> => {
  let sortByQuery: string | undefined = undefined;
  if (sortBy) {
    let field;
    switch (sortBy.field) {
      case BusinessServiceSortBy.NAME:
        field = "name";
        break;
      case BusinessServiceSortBy.OWNER:
        field = "owner.displayName";
        break;
      default:
        throw new Error("Could not define SortBy field name");
    }
    sortByQuery = `${sortBy.direction === "desc" ? "-" : ""}${field}`;
  }

  const params = {
    page: pagination.page - 1,
    size: pagination.perPage,
    sort: sortByQuery,

    name: filters.name,
    description: filters.description,
    "owner.displayName": filters.owner,
  };

  const query: string[] = buildQuery(params);
  return APIClient.get(`${BUSINESS_SERVICES}?${query.join("&")}`, { headers });
};

export const deleteBusinessService = (id: number | string): AxiosPromise => {
  return APIClient.delete(`${BUSINESS_SERVICES}/${id}`);
};

export const createBusinessService = (
  obj: BusinessService
): AxiosPromise<BusinessService> => {
  return APIClient.post(`${BUSINESS_SERVICES}`, obj);
};

export const updateBusinessService = (
  obj: BusinessService
): AxiosPromise<BusinessService> => {
  return APIClient.put(`${BUSINESS_SERVICES}/${obj.id}`, obj);
};

export const getBusinessServiceById = (
  id: number | string
): AxiosPromise<BusinessService> => {
  return APIClient.get(`${BUSINESS_SERVICES}/${id}`);
};

// Stakeholders

export enum StakeholderSortBy {
  EMAIL,
  DISPLAY_NAME,
  JOB_FUNCTION,
  STAKEHOLDER_GROUPS,
}
export interface StakeholderSortByQuery {
  field: StakeholderSortBy;
  direction?: Direction;
}

export const getStakeholders = (
  filters: {
    email?: string[];
    displayName?: string[];
    jobFunction?: string[];
    stakeholderGroup?: string[];
  },
  pagination: PageQuery,
  sortBy?: StakeholderSortByQuery
): AxiosPromise<StakeholderPage> => {
  let sortByQuery: string | undefined = undefined;
  if (sortBy) {
    let field;
    switch (sortBy.field) {
      case StakeholderSortBy.EMAIL:
        field = "email";
        break;
      case StakeholderSortBy.DISPLAY_NAME:
        field = "displayName";
        break;
      case StakeholderSortBy.JOB_FUNCTION:
        field = "jobFunction.role";
        break;
      case StakeholderSortBy.STAKEHOLDER_GROUPS:
        field = "stakeholderGroups.size()";
        break;
      default:
        throw new Error("Could not define SortBy field name");
    }
    sortByQuery = `${sortBy.direction === "desc" ? "-" : ""}${field}`;
  }

  const params = {
    page: pagination.page - 1,
    size: pagination.perPage,
    sort: sortByQuery,

    email: filters.email,
    displayName: filters.displayName,
    "jobFunction.role": filters.jobFunction,
    "stakeholderGroups.name": filters.stakeholderGroup,
  };

  const query: string[] = buildQuery(params);
  return APIClient.get(`${STAKEHOLDERS}?${query.join("&")}`, { headers });
};

export const createJobFunction = (
  obj: JobFunction
): AxiosPromise<JobFunction> => {
  return APIClient.post(`${JOB_FUNCTIONS}`, obj);
};

export const updateJobFunction = (
  obj: JobFunction
): AxiosPromise<JobFunction> => {
  return APIClient.put(`${JOB_FUNCTIONS}/${obj.id}`, obj);
};

export const deleteJobFunction = (id: number): AxiosPromise => {
  return APIClient.delete(`${JOB_FUNCTIONS}/${id}`);
};

export const deleteStakeholder = (id: number): AxiosPromise => {
  return APIClient.delete(`${STAKEHOLDERS}/${id}`);
};

export const createStakeholder = (
  obj: Stakeholder
): AxiosPromise<Stakeholder> => {
  return APIClient.post(`${STAKEHOLDERS}`, obj);
};

export const updateStakeholder = (
  obj: Stakeholder
): AxiosPromise<Stakeholder> => {
  return APIClient.put(`${STAKEHOLDERS}/${obj.id}`, obj);
};

// Stakeholder groups

export enum StakeholderGroupSortBy {
  NAME,
  STAKEHOLDERS,
}
export interface StakeholderGroupSortByQuery {
  field: StakeholderGroupSortBy;
  direction?: Direction;
}

export const getStakeholderGroups = (
  filters: {
    name?: string[];
    description?: string[];
    stakeholder?: string[];
  },
  pagination: PageQuery,
  sortBy?: StakeholderGroupSortByQuery
): AxiosPromise<StakeholderGroupPage> => {
  let sortByQuery: string | undefined = undefined;
  if (sortBy) {
    let field;
    switch (sortBy.field) {
      case StakeholderGroupSortBy.NAME:
        field = "name";
        break;
      case StakeholderGroupSortBy.STAKEHOLDERS:
        field = "stakeholders.size()";
        break;
      default:
        throw new Error("Could not define SortBy field name");
    }
    sortByQuery = `${sortBy.direction === "desc" ? "-" : ""}${field}`;
  }

  const params = {
    page: pagination.page - 1,
    size: pagination.perPage,
    sort: sortByQuery,

    name: filters.name,
    description: filters.description,
    "stakeholders.displayName": filters.stakeholder,
  };

  const query: string[] = buildQuery(params);
  return APIClient.get(`${STAKEHOLDER_GROUPS}?${query.join("&")}`, { headers });
};

export const deleteStakeholderGroup = (id: number): AxiosPromise => {
  return APIClient.delete(`${STAKEHOLDER_GROUPS}/${id}`);
};

export const createStakeholderGroup = (
  obj: StakeholderGroup
): AxiosPromise<StakeholderGroup> => {
  return APIClient.post(`${STAKEHOLDER_GROUPS}`, obj);
};

export const updateStakeholderGroup = (
  obj: StakeholderGroup
): AxiosPromise<StakeholderGroup> => {
  return APIClient.put(`${STAKEHOLDER_GROUPS}/${obj.id}`, obj);
};

// Job functions

export enum JobFunctionSortBy {
  ROLE,
}
export interface JobFunctionSortByQuery {
  field: JobFunctionSortBy;
  direction?: Direction;
}

export const getJobFunctions = (
  filters: {
    role?: string[];
  },
  pagination: PageQuery,
  sortBy?: JobFunctionSortByQuery
): AxiosPromise<JobFunctionPage> => {
  let sortByQuery: string | undefined = undefined;
  if (sortBy) {
    let field;
    switch (sortBy.field) {
      case JobFunctionSortBy.ROLE:
        field = "role";
        break;
      default:
        throw new Error("Could not define SortBy field name");
    }
    sortByQuery = `${sortBy.direction === "desc" ? "-" : ""}${field}`;
  }

  const params = {
    page: pagination.page - 1,
    size: pagination.perPage,
    sort: sortByQuery,

    role: filters.role,
  };

  const query: string[] = buildQuery(params);
  return APIClient.get(`${JOB_FUNCTIONS}?${query.join("&")}`, { headers });
};

// App inventory

export enum ApplicationSortBy {
  NAME,
}
export interface ApplicationSortByQuery {
  field: ApplicationSortBy;
  direction?: Direction;
}

export const getApplications = (
  filters: {
    name?: string[];
    description?: string[];
    businessService?: string[];
  },
  pagination: PageQuery,
  sortBy?: ApplicationSortByQuery
): AxiosPromise<ApplicationPage> => {
  let sortByQuery: string | undefined = undefined;
  if (sortBy) {
    let field;
    switch (sortBy.field) {
      case ApplicationSortBy.NAME:
        field = "name";
        break;
      default:
        throw new Error("Could not define SortBy field name");
    }
    sortByQuery = `${sortBy.direction === "desc" ? "-" : ""}${field}`;
  }

  const params = {
    page: pagination.page - 1,
    size: pagination.perPage,
    sort: sortByQuery,

    name: filters.name,
    description: filters.description,
    businessService: filters.businessService,
  };

  const query: string[] = buildQuery(params);
  return APIClient.get(`${APPLICATIONS}?${query.join("&")}`, {
    headers,
  });
};

export const deleteApplication = (id: number): AxiosPromise => {
  return APIClient.delete(`${APPLICATIONS}/${id}`);
};

export const createApplication = (
  obj: Application
): AxiosPromise<Application> => {
  return APIClient.post(`${APPLICATIONS}`, obj);
};

export const updateApplication = (
  obj: Application
): AxiosPromise<Application> => {
  return APIClient.put(`${APPLICATIONS}/${obj.id}`, obj);
};

export const getApplicationById = (
  id: number | string
): AxiosPromise<Application> => {
  return APIClient.get(`${APPLICATIONS}/${id}`);
};

//

export const getApplicationDependencies = (
  filters: {
    from?: string[];
    to?: string[];
  },
  pagination: PageQuery
): AxiosPromise<ApplicationDependencyPage> => {
  const params = {
    page: pagination.page - 1,
    size: pagination.perPage,

    "from.id": filters.from,
    "to.id": filters.to,
  };

  const query: string[] = buildQuery(params);
  return APIClient.get(`${APPLICATION_DEPENDENCY}?${query.join("&")}`, {
    headers,
  });
};

export const createApplicationDependency = (
  obj: ApplicationDependency
): AxiosPromise<ApplicationDependency> => {
  return APIClient.post(`${APPLICATION_DEPENDENCY}`, obj);
};

export const deleteApplicationDependency = (id: number): AxiosPromise => {
  return APIClient.delete(`${APPLICATION_DEPENDENCY}/${id}`);
};

//

export const getAssessments = (filters: {
  applicationId?: number;
}): AxiosPromise<Assessment[]> => {
  const params = {
    applicationId: filters.applicationId,
  };

  const query: string[] = buildQuery(params);
  return APIClient.get(`${ASSESSMENTS}?${query.join("&")}`);
};

export const createAssessment = (obj: Assessment): AxiosPromise<Assessment> => {
  return APIClient.post(`${ASSESSMENTS}`, obj);
};

export const getAssessmentById = (
  id: number | string
): AxiosPromise<Assessment> => {
  return APIClient.get(`${ASSESSMENTS}/${id}`);
};
