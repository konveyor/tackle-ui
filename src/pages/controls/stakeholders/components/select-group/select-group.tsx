import React, { useCallback, useState } from "react";
import { AxiosError } from "axios";

import {
  Select,
  SelectOption,
  SelectOptionObject,
  SelectVariant,
} from "@patternfly/react-core";
import { WarningTriangleIcon } from "@patternfly/react-icons";

import { StakeholderGroup } from "api/models";
import { getAxiosErrorMessage } from "utils/utils";

interface SelectOptionStakeholder extends SelectOptionObject {
  stakeholderGroup: StakeholderGroup;
}

const selectOptionMapper = (
  stakeholderGroup: StakeholderGroup
): SelectOptionStakeholder => ({
  stakeholderGroup: { ...stakeholderGroup },
  toString: function () {
    return stakeholderGroup.name;
  },
});

export interface SelectGroupProps {
  value?: StakeholderGroup[];
  stakeholderGroups: StakeholderGroup[];
  isFetching?: boolean;
  fetchError?: AxiosError;
  onSelect: (value: StakeholderGroup[]) => void;
  onClear: () => void;
}

export const SelectGroup: React.FC<SelectGroupProps> = ({
  value,
  stakeholderGroups,
  isFetching,
  fetchError,
  onSelect,
  onClear,
}) => {
  const [isOpen, setIsOpen] = useState(false);

  const handleOnToggle = useCallback((isOpen: boolean) => {
    setIsOpen(isOpen);
  }, []);

  const handleOnSelect = (
    event: React.MouseEvent | React.ChangeEvent,
    selection: string | SelectOptionObject
  ) => {
    if (typeof selection === "string") {
      throw new Error("This selection is not allowed");
    }

    setIsOpen(false);

    const selectedOption = selection as SelectOptionStakeholder;

    if (value?.find((f) => f.id === selectedOption.stakeholderGroup.id)) {
      onSelect(
        value.filter((f) => f.id !== selectedOption.stakeholderGroup.id)
      );
    } else {
      onSelect([...(value || []), selectedOption.stakeholderGroup]);
    }
  };

  const handleOnClearSelection = () => {
    onClear();
  };

  let customContent;
  if (isFetching) {
    customContent = <div>&nbsp;Loading...</div>;
  } else if (fetchError) {
    customContent = <div>&nbsp;{getAxiosErrorMessage(fetchError)}</div>;
  }

  return (
    <Select
      toggleIcon={fetchError && <WarningTriangleIcon />}
      variant={SelectVariant.typeaheadMulti}
      onToggle={handleOnToggle}
      onSelect={handleOnSelect}
      onClear={handleOnClearSelection}
      selections={value?.map((f) => selectOptionMapper(f))}
      isOpen={isOpen}
      menuAppendTo={() => document.body}
      maxHeight={350}
      customContent={customContent}
      placeholderText="Select a stakeholder group"
    >
      {stakeholderGroups
        .map((f) => selectOptionMapper(f))
        .map((elem, index) => (
          <SelectOption key={index} value={elem} />
        ))}
    </Select>
  );
};
