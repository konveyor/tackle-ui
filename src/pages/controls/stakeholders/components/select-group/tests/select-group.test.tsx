import React from "react";
import { shallow } from "enzyme";

import { StakeholderGroup } from "api/models";
import { SelectGroup } from "../select-group";

describe("SelectGroup", () => {
  const STAKEHOLDER_GROUPS: StakeholderGroup[] = [
    {
      name: "aa",
      description: "somethingA",
    },
    {
      name: "bb",
      description: "somethingB",
    },
  ];

  it("Renders without crashing", () => {
    const wrapper = shallow(
      <SelectGroup
        stakeholderGroups={STAKEHOLDER_GROUPS}
        onSelect={jest.fn()}
        onClear={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
