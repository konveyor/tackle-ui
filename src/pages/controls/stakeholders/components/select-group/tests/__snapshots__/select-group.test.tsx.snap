// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectGroup Renders without crashing 1`] = `
<Select
  aria-label=""
  aria-labelledby=""
  chipGroupComponent={null}
  className=""
  clearSelectionsAriaLabel="Clear all"
  createText="Create"
  customBadgeText={null}
  customContent={null}
  direction="down"
  favorites={Array []}
  favoritesLabel="Favorites"
  hasInlineFilter={false}
  inlineFilterPlaceholderText={null}
  inputIdPrefix=""
  isCreatable={false}
  isDisabled={false}
  isGrouped={false}
  isOpen={false}
  isPlain={false}
  maxHeight={350}
  menuAppendTo={[Function]}
  noResultsFoundText="No results found"
  onClear={[Function]}
  onCreateOption={[Function]}
  onFilter={null}
  onSelect={[Function]}
  onToggle={[Function]}
  onTypeaheadInputChanged={null}
  ouiaSafe={true}
  placeholderText="Select stakeholder groups"
  removeSelectionAriaLabel="Remove"
  selections={Array []}
  toggleAriaLabel="Options menu"
  toggleIcon={null}
  toggleId={null}
  typeAheadAriaLabel=""
  variant="typeaheadmulti"
  width=""
>
  <SelectOption
    className=""
    component="button"
    index={0}
    inputId=""
    isChecked={false}
    isDisabled={false}
    isFavorite={null}
    isNoResultsOption={false}
    isPlaceholder={false}
    isSelected={false}
    key="0"
    keyHandler={[Function]}
    onClick={[Function]}
    sendRef={[Function]}
    value={
      Object {
        "stakeholderGroup": Object {
          "description": "somethingA",
          "name": "aa",
        },
        "toString": [Function],
      }
    }
  />
  <SelectOption
    className=""
    component="button"
    index={0}
    inputId=""
    isChecked={false}
    isDisabled={false}
    isFavorite={null}
    isNoResultsOption={false}
    isPlaceholder={false}
    isSelected={false}
    key="1"
    keyHandler={[Function]}
    onClick={[Function]}
    sendRef={[Function]}
    value={
      Object {
        "stakeholderGroup": Object {
          "description": "somethingB",
          "name": "bb",
        },
        "toString": [Function],
      }
    }
  />
</Select>
`;
